File: content_agent/fake.py
"""Client giả lập cho --dry-run: chạy hết pipeline không gọi API, không tốn tiền.

Sinh dữ liệu bám đúng JSON Schema được gửi lên, nên nó kiểm tra được cả phần
schema lẫn phần ghép file — chỉ nội dung là vô nghĩa.
"""

from __future__ import annotations

import json
from types import SimpleNamespace
from typing import Any


def _sample_for(schema: dict, key: str = "") -> Any:
    kind = schema.get("type")
    if "enum" in schema:
        return schema["enum"][0]
    if kind == "object":
        return {k: _sample_for(v, k) for k, v in schema.get("properties", {}).items()}
    if kind == "array":
        return [_sample_for(schema.get("items", {"type": "string"}), key) for _ in range(3)]
    if kind == "integer":
        return 20 if "duration" in key else 3
    if kind == "number":
        return 1.0
    if kind == "boolean":
        return True
    if "url" in key:
        return "https://example.com/nguon-mau"
    return f"[dry-run] nội dung mẫu cho {key or 'trường này'}"


class FakeMessages:
    def create(self, **params):
        config = params.get("output_config") or params.get("extra_body", {}).get(
            "output_config", {}
        )
        schema = (config.get("format") or {}).get("schema")
        if schema:
            text = json.dumps(_sample_for(schema), ensure_ascii=False)
        else:
            text = (
                "[dry-run] Bản ghi chú nghiên cứu mẫu.\n"
                "- [độ tin cậy: cao] Một thông tin mẫu — Nguồn: https://example.com/a\n"
                "- [độ tin cậy: thấp] Một thông tin chưa chắc — Nguồn: https://example.com/b"
            )
        return SimpleNamespace(
            content=[SimpleNamespace(type="text", text=text)],
            stop_reason="end_turn",
            stop_details=None,
            usage=SimpleNamespace(input_tokens=0, output_tokens=0),
        )


class FakeClient:
    def __init__(self) -> None:
        self.messages = FakeMessages()

File: src/llm/fake.ts
// Dry-run backend. Generates data conforming to whatever JSON Schema the request
// carries, so `dry_run: true` exercises the real schema + packaging path with
// zero API calls and zero cost.

import type { CreateParams, LlmResponse } from "./types";

function sampleFor(schema: any, key = ""): unknown {
  if (schema && Array.isArray(schema.enum)) return schema.enum[0];
  if (schema?.type === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(schema.properties || {})) {
      out[k] = sampleFor(v, k);
    }
    return out;
  }
  if (schema?.type === "array") {
    const itemSchema = schema.items || { type: "string" };
    return [0, 1, 2].map(() => sampleFor(itemSchema, key));
  }
  if (schema?.type === "integer") {
    return key.includes("duration") ? 20 : 3;
  }
  if (schema?.type === "number") return 1.0;
  if (schema?.type === "boolean") return true;
  if (key.includes("url")) return "https://example.com/nguon-mau";
  return `[dry-run] nội dung mẫu cho ${key || "trường này"}`;
}

const FALLBACK_TEXT =
  "- [độ tin cậy: cao] Đây là thông tin mẫu cho chế độ chạy thử — Nguồn: https://example.com/nguon-mau\n" +
  "- [độ tin cậy: thấp] Thông tin mẫu chưa kiểm chứng — Nguồn: https://example.com/nguon-mau-2";

export async function callFake(params: CreateParams): Promise<LlmResponse> {
  const schema = params.outputConfig?.format?.schema;
  const text = schema ? JSON.stringify(sampleFor(schema)) : FALLBACK_TEXT;
  return {
    content: [{ type: "text", text }],
    stopReason: "end_turn",
    stopDetails: null,
    usage: { inputTokens: 0, outputTokens: 0, costUsd: null },
  };
}
